File: backend/app/services/inventory.py
import hashlib
import mimetypes
from pathlib import Path
from uuid import uuid4

from PIL import Image
from pypdf import PdfReader

from app.models.schemas import AuditResponse, FileAudit
from app.services.costs import estimate_processing_cost
from app.services.paths import resolve_source_path

SUPPORTED_EXTENSIONS = {".pdf", ".png", ".jpg", ".jpeg", ".tif", ".tiff", ".webp"}


def discover_supported_files(source_path: str, limit: int | None = None) -> list[Path]:
    root = resolve_source_path(source_path)
    if not root.exists():
        raise FileNotFoundError(source_path)

    files = [root] if root.is_file() else [path for path in root.rglob("*") if path.is_file()]
    supported_files = [path for path in files if path.suffix.lower() in SUPPORTED_EXTENSIONS]
    supported_files.sort(key=lambda path: str(path).lower())
    return supported_files[:limit] if limit else supported_files


def audit_source(source_path: str, sample_limit: int) -> AuditResponse:
    if sample_limit < 1:
        raise ValueError("sample_limit must be positive")

    files = discover_supported_files(source_path)
    sampled = files[:sample_limit]
    audits = [_audit_file(path) for path in sampled]
    total_bytes = sum(path.stat().st_size for path in files)
    total_pages = _estimate_total_pages(files, audits)
    scanned_pages = _estimate_scanned_pages(files, audits, total_pages)
    native_text_pages = max(total_pages - scanned_pages, 0)

    return AuditResponse(
        run_id=uuid4(),
        source_path=source_path,
        total_files=len(files),
        total_bytes=total_bytes,
        total_pages=total_pages,
        scanned_pages=scanned_pages,
        native_text_pages=native_text_pages,
        sampled_files=audits,
        estimate=estimate_processing_cost(total_pages, scanned_pages),
    )


def extract_text_by_page(file_path: Path) -> list[str]:
    if file_path.suffix.lower() != ".pdf":
        return [""]

    reader = PdfReader(str(file_path))
    return [(page.extract_text() or "").strip() for page in reader.pages]


def _audit_file(path: Path) -> FileAudit:
    mime_type = mimetypes.guess_type(path.name)[0] or "application/octet-stream"
    page_count, has_native_text = _inspect_document(path)
    is_probably_scanned = not has_native_text

    return FileAudit(
        path=str(path),
        filename=path.name,
        extension=path.suffix.lower(),
        mime_type=mime_type,
        size_bytes=path.stat().st_size,
        sha256=_sha256(path),
        page_count=page_count,
        has_native_text=has_native_text,
        is_probably_scanned=is_probably_scanned,
    )


def _inspect_document(path: Path) -> tuple[int, bool]:
    if path.suffix.lower() == ".pdf":
        return _inspect_pdf(path)
    return _inspect_image(path)


def _inspect_pdf(path: Path) -> tuple[int, bool]:
    reader = PdfReader(str(path))
    page_count = len(reader.pages)
    text_pages = 0
    for page in reader.pages[:5]:
        text = (page.extract_text() or "").strip()
        if len(text) >= 40:
            text_pages += 1
    return page_count, text_pages > 0


def _inspect_image(path: Path) -> tuple[int, bool]:
    with Image.open(path) as image:
        image.verify()
    return 1, False


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as file:
        for chunk in iter(lambda: file.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def _estimate_total_pages(files: list[Path], audits: list[FileAudit]) -> int:
    if len(files) == len(audits):
        return sum(audit.page_count for audit in audits)
    if not audits:
        return 0
    average_pages = sum(audit.page_count for audit in audits) / len(audits)
    return round(average_pages * len(files))


def _estimate_scanned_pages(files: list[Path], audits: list[FileAudit], total_pages: int) -> int:
    if not audits or total_pages == 0:
        return 0
    sampled_pages = sum(audit.page_count for audit in audits)
    scanned_sample_pages = sum(audit.page_count for audit in audits if audit.is_probably_scanned)
    if len(files) == len(audits):
        return scanned_sample_pages
    scanned_ratio = scanned_sample_pages / sampled_pages if sampled_pages else 0
    return round(total_pages * scanned_ratio)

File: internal/localengine/inventory/inventory.go
package inventory

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"mime"
	"os"
	"path/filepath"
	"strings"

	"github.com/google/uuid"

	"archivo-digital-inteligente/internal/localengine/models"
	"archivo-digital-inteligente/internal/localengine/pdf"
)

var supportedExtensions = map[string]struct{}{
	".pdf":  {},
	".png":  {},
	".jpg":  {},
	".jpeg": {},
	".tif":  {},
	".tiff": {},
	".webp": {},
}

func DiscoverSupportedFiles(sourcePath string, limit *int) ([]string, error) {
	root := filepath.Clean(sourcePath)
	info, err := os.Stat(root)
	if err != nil {
		return nil, fmt.Errorf("source path not found: %w", err)
	}
	var files []string
	if !info.IsDir() {
		if isSupported(root) {
			files = append(files, root)
		}
	} else {
		err = filepath.WalkDir(root, func(path string, entry os.DirEntry, walkErr error) error {
			if walkErr != nil {
				return walkErr
			}
			if entry.IsDir() {
				return nil
			}
			if isSupported(path) {
				files = append(files, path)
			}
			return nil
		})
		if err != nil {
			return nil, err
		}
	}
	if limit != nil && len(files) > *limit {
		files = files[:*limit]
	}
	return files, nil
}

func AuditSource(sourcePath string, sampleLimit int) (models.AuditResponse, error) {
	if sampleLimit < 1 {
		return models.AuditResponse{}, fmt.Errorf("sample_limit must be positive")
	}
	files, err := DiscoverSupportedFiles(sourcePath, nil)
	if err != nil {
		return models.AuditResponse{}, err
	}
	sampledCount := sampleLimit
	if len(files) < sampledCount {
		sampledCount = len(files)
	}
	sampled := make([]models.FileAudit, 0, sampledCount)
	for _, path := range files[:sampledCount] {
		audit, auditErr := auditFile(path)
		if auditErr != nil {
			continue
		}
		sampled = append(sampled, audit)
	}
	var totalBytes int64
	for _, path := range files {
		info, statErr := os.Stat(path)
		if statErr == nil {
			totalBytes += info.Size()
		}
	}
	totalPages := estimateTotalPages(len(files), sampled)
	scannedPages := estimateScannedPages(len(files), sampled, totalPages)
	return models.AuditResponse{
		RunID:           uuid.NewString(),
		SourcePath:      sourcePath,
		TotalFiles:      len(files),
		TotalBytes:      totalBytes,
		TotalPages:      totalPages,
		ScannedPages:    scannedPages,
		NativeTextPages: max(totalPages-scannedPages, 0),
		SampledFiles:    sampled,
		Estimate:        estimateProcessingCost(totalPages, scannedPages),
	}, nil
}

func auditFile(path string) (models.FileAudit, error) {
	info, err := os.Stat(path)
	if err != nil {
		return models.FileAudit{}, err
	}
	ext := strings.ToLower(filepath.Ext(path))
	mimeType := mime.TypeByExtension(ext)
	if mimeType == "" {
		mimeType = "application/octet-stream"
	}
	pageCount := 1
	hasNativeText := false
	isProbablyScanned := true
	if ext == ".pdf" {
		var inspectErr error
		pageCount, hasNativeText, inspectErr = pdf.InspectPDF(path)
		if inspectErr != nil {
			pageCount = 1
			hasNativeText = false
			isProbablyScanned = true
		} else {
			isProbablyScanned = !hasNativeText
		}
	}
	hash, err := sha256File(path)
	if err != nil {
		return models.FileAudit{}, err
	}
	return models.FileAudit{
		Path:              path,
		Filename:          filepath.Base(path),
		Extension:         ext,
		MimeType:          mimeType,
		SizeBytes:         info.Size(),
		SHA256:            hash,
		PageCount:         pageCount,
		HasNativeText:     hasNativeText,
		IsProbablyScanned: isProbablyScanned,
	}, nil
}

func sha256File(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()
	hasher := sha256.New()
	if _, err := io.Copy(hasher, file); err != nil {
		return "", err
	}
	return hex.EncodeToString(hasher.Sum(nil)), nil
}

func estimateTotalPages(totalFiles int, sampled []models.FileAudit) int {
	if totalFiles == len(sampled) {
		total := 0
		for _, audit := range sampled {
			total += audit.PageCount
		}
		return total
	}
	if len(sampled) == 0 {
		return 0
	}
	sum := 0
	for _, audit := range sampled {
		sum += audit.PageCount
	}
	average := float64(sum) / float64(len(sampled))
	return int(average*float64(totalFiles) + 0.5)
}

func estimateScannedPages(totalFiles int, sampled []models.FileAudit, totalPages int) int {
	if len(sampled) == 0 || totalPages == 0 {
		return 0
	}
	if totalFiles == len(sampled) {
		scanned := 0
		for _, audit := range sampled {
			if audit.IsProbablyScanned {
				scanned += audit.PageCount
			}
		}
		return scanned
	}
	sampledPages := 0
	scannedSample := 0
	for _, audit := range sampled {
		sampledPages += audit.PageCount
		if audit.IsProbablyScanned {
			scannedSample += audit.PageCount
		}
	}
	if sampledPages == 0 {
		return 0
	}
	ratio := float64(scannedSample) / float64(sampledPages)
	return int(ratio*float64(totalPages) + 0.5)
}

func estimateProcessingCost(totalPages int, scannedPages int) models.CostEstimate {
	geminiExtract := float64(totalPages) * 0.0008
	geminiEmbed := float64(totalPages) * 0.0002
	fallbackLow := float64(totalPages) * 0.001
	fallbackHigh := float64(totalPages) * 0.003
	totalLow := geminiExtract + geminiEmbed + fallbackLow
	totalHigh := geminiExtract + geminiEmbed + fallbackHigh
	return models.CostEstimate{
		Pages:                    totalPages,
		ScannedPages:             scannedPages,
		NativeTextPages:          max(totalPages-scannedPages, 0),
		GoogleOCRUSD:             0,
		GeminiExtractionUSD:      geminiExtract,
		GeminiEmbeddingUSD:       geminiEmbed,
		AnthropicFallbackLowUSD:  fallbackLow,
		AnthropicFallbackHighUSD: fallbackHigh,
		TotalLowUSD:              totalLow,
		TotalHighUSD:             totalHigh,
	}
}

func isSupported(path string) bool {
	_, ok := supportedExtensions[strings.ToLower(filepath.Ext(path))]
	return ok
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
